cloud-checksum/src/task/copy.rs

```diff
@@ -52,17 +96,214 @@ impl CopyTaskBuilder {
+                .build(self.source.to_string())
+                .await?
+                .sums_file()
+                .await?
+        } else {
+            None
+        };
+
+        // If there are existing sums, try the best part size.
+        let settings = if let Some(sums) = sums {
+            let settings = self.use_settings_from_sums(
+                &sums,
+                ObjectInfo {
+                    size,
+                    max_parts,
+                    max_part_size,
+                    min_part_size,
+                },
+                destination,
+            )?;
+            if self.part_size.is_none() {
+                return Ok(settings);
+            } else {
+                Some(settings)
+            }
+        } else {
+            None
+        };
+
+        // Use the additional sum from the settings if available or the default.
+        let additional_ctx = settings
+            .map(|settings| settings.into_inner().1)
+            .unwrap_or_default();
+
+        let threshold = self
+            .multipart_threshold
+            .unwrap_or(DEFAULT_MULTIPART_THRESHOLD);
+
+        // If the part size is set, use that.
+        if let Some(part_size) = self.part_size {
+            if size > threshold {
+                return if Self::is_multipart(
+                    size,
+                    part_size,
+                    max_parts,
+                    max_part_size,
+                    min_part_size,
+                ) {
+                    Ok(CopySettings::new(Some(part_size), additional_ctx, size))
+                } else {
+                    Err(CopyError(format!(
+                        "invalid part size `{}` and threshold `{}` for the object size `{}`",
+                        part_size, threshold, size
+                    )))
+                };
+            }
+        }
+
+        let err = || {
+            CopyError(format!(
+                "failed to find a valid part size for the threshold `{}` with object size `{}`",
+                threshold, size
+            ))
+        };
+        // Use multipart if the size reaches the threshold.
+        if size > threshold {
+            let part_size = PART_SIZE_ORDERING.keys().copied().find(|part_size| {
+                Self::is_multipart(size, *part_size, max_parts, max_part_size, min_part_size)
+            });
+
+            return if let Some(part_size) = part_size {
+                Ok(CopySettings::new(Some(part_size), additional_ctx, size))
+            } else {
+                Err(err())
+            };
+        }
+
+        // Otherwise use single part if possible.
+        if Self::is_single_part(size, max_part_size) {
+            return Ok(CopySettings::new(None, additional_ctx, size));
+        }
+
+        // This condition may occur if the size is greater than the possible single part upload
+        // limit but lower than the threshold.
+        Err(err())
+    }
```

Comment:
This is the function that computes the "preferred" part size and determines whether to use multipart/single part uploads. Some of these choices are arbitrary but I think it's okay as long as the "check" afterwards doesn't have to compute checksums if possible.

Reply:
yeah agreed there's going to be some arbitrary logic in here. Happy to go with anything at the moment and iterate.